### DISCUSSION TITLE
morph: removing / removed not called for keyed element despite it being removed

### DISCUSSION THREAD

**Discussion by SteffenDE** [alpinejs/alpine#4034] [7. Bugs (archived)]
I am currently playing with the morph plugin and am a little bit confused because a node is removed (as it should!), but the removing / removed hook is not called:

```html
<html>
	<head>
		<!-- Alpine Plugins -->
		<script defer src="https://cdn.jsdelivr.net/npm/@alpinejs/morph@3.13.5/dist/cdn.min.js"></script>
		<!-- Alpine Core -->
		<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.5/dist/cdn.min.js"></script>
	</head>
	<body>
		<ul id="container">
			<li key="1">Item 1</li>
			<li key="2">Item 2</div>
		</ul>
	</body>

	<script defer>
		document.addEventListener('alpine:initialized', () => {
			window.Alpine.morph(document.getElementById("container"), `
			<ul id="container">
				<li key="1">Item 1</li>
				<li key="3">Item 3</div>
			</ul>
		`, {
				key: (el) => el.hasAttribute && el.getAttribute("key"),
				adding: (el) => console.log("adding", el),
				removing: (el) => console.log("removing", el),
				updating: (fromEl, toEl, _children_only, _skip) => console.log("updating", fromEl, toEl)
			})
		})
	</script>
</html>
```

The id of the elements are used as keys. In the logs you can see `Item 3` added, but `Item 2` is not passed to removing, despite it being removed.

**Comment by ekwoka** [comment]
Hmm, that is a strange one. I did verify this is happening, but I'm not sure exactly why. The code certainly SEEMS like it should handle it fine, and it does seem to be seeing and keying the old not properly.

Lookahead also doesn't change anything (tested to see if maybe all the text nodes were a factor)

**Comment by SteffenDE** [comment]
I adjusted the initial HTML to be simpler.

If I remove this line
https://github.com/alpinejs/alpine/blob/ec11d59e22136a9a9c45895ba5a13967695895f8/packages/morph/src/morph.js#L275

It seems to work fine. Maybe
https://github.com/alpinejs/alpine/blob/ec11d59e22136a9a9c45895ba5a13967695895f8/packages/morph/src/morph.js#L248
as well.

I'm not sure if this has other consequences.

**Reply by SteffenDE** [reply]
(at least the tests still pass with those changes)

**Comment by SteffenDE** [comment]
So... another case where morph removes keyed elements:

```html
<html>
  <head>
    <!-- Alpine Plugins -->
    <script defer src="https://cdn.jsdelivr.net/npm/@alpinejs/morph@3.13.5/dist/cdn.min.js"></script>
    <!-- Alpine Core -->
    <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.5/dist/cdn.min.js"></script>
  </head>
  <body>
    <div id="container">
      <div id="list" data-keep-children>
        <div id="2">2</div>
        <div id="3">3</div>
        <div id="1">1</div>
      </div>
    </div>
  </body>

  <script defer>
    document.addEventListener('alpine:initialized', () => {
      window.Alpine.morph(document.getElementById("container"), `
      <div id="container">
        <div id="list" data-keep-children>
          <div id="3">Updated element</div>
        </div>
      </div>
    `, {
        key: (el) => el.hasAttribute && el.getAttribute("id"),
        adding: (el) => console.log("adding", el),
        removing: (el, skip) => {
          if(el.parentElement !== null && el.id && el.parentElement.getAttribute("data-keep-children")!= null){
            console.log("not removing", el.id)
            return skip()
          }
        },
        updating: (fromEl, toEl, _childrenOnly, _skip) => {
          console.log("updating", fromEl, toEl)
        }
      })
    })
  </script>
</html>
```

Here there is a container with special attribute `data-keep-children` that is checked in removing. But even if I apply the changes from the previous comments, the element with id `2` is removed in this line:

https://github.com/alpinejs/alpine/blob/ec11d59e22136a9a9c45895ba5a13967695895f8/packages/morph/src/morph.js#L269

removing is not called and therefore there's no chance to skip the removal.

**Comment by SteffenDE** [comment]
https://github.com/alpinejs/alpine/compare/main...SteffenDE:alpine:morph_remove_bug
